_posts/2018/2018-01-07-extract-conditionals-into-well-named-methods.md

```diff
@@ -0,0 +1,97 @@
+---
+title: 'Extract conditionals into well-named methods'
+description: 'An extraction to a private method often adds clarity'
+layout: article
+category: ruby
+image:
+  base: '2018/extract-conditionals-into-well-named-methods'
+  alt: 'Jigsaw pieces'
+  source: 'https://unsplash.com/photos/3y1zF4hIPCg'
+  credit: 'Hans-Peter Gauster'
+---
+
+One of the simplest (and most effective) ways to refactor is to [extract a method](https://refactoring.com/catalog/extractMethod.html). The naming of an extracted method, can be especially good for communicating about thinking behind the code.
+
+## Instead of…
+
+...using slightly complex statements in your conditionals.
+
+```ruby
+class BrightonCoffeeShop
+  def initialize(name)
+    @name = name
+  end
+
+  def good?
+    if @name == 'Starbucks' || @name == 'Costa' ||
+       @name == 'Barry’s Caff'
+      false
+    elsif @name == 'Coffee@33' || @name == 'Small Batch'
+      true
+    else
+      false
+    end
+  end
+end
+```
+
+This implementation explicitly returns `false` if the `name` is unknown. This is the only way to guarantee a decent coffee shop experience.
+
+
+## Use…
+
+...the extract method refactoring to name the concept and move it to a `private` method.
+
+```ruby
+class BrightonCoffeeShop
+  def initialize(name)
+    @name = name
+  end
+
+  def good?
+    clean? && local? && excellent_coffee?
+  end
+
+  private
+
+  def clean?
+    !filthy?
+  end
+
+  def excellent_coffee?
+    ['Coffee@33', 'Small Batch'].include?(@name)
+  end
+
+  def filthy?
+    @name == 'Barry’s Caff'
+  end
+
+  def local?
+    !national_chain?
+  end
+
+  def national_chain?
+    ['Starbucks', 'Costa'].include?(@name)
+  end
+end
+```
+
+We've also refactored to use `Array#include?` for the lists of names.
+
+
+## But why?
+
+Here, by extracting and naming methods, we have better described what exactly makes a good coffee shop. The private methods act as clear documentation for the class. Before all we really had was a list of 'good' and 'bad' names. You can tell what makes a good coffee shop without even referring to the content of the private methods.
```

Comment:
*Here, by extracting methods...

By extracting a method you, by definition, name it.

Reply:
*...better described what makes a good coffee shop

OR

* described the specific criteria for a good coffee shop

OR

* described exactly what makes a good coffee shop

I think my first suggestion is the best.

Reply:
* Before, all we really had was a list of 'good' and 'bad' coffee shops. _We had no knowledge of how those lists were developed._ (Or something like that) Now, you can tell what makes a good coffee shop without even having to refer to the content of the extracted methods.